[CISupport][build.webkit.org] build step clean-git-index-lock is not working

clopez opened the pull request:
#### f3acc6a57d57923fe2b62fe2b8af6489d5d4f1fd
<pre>
[CISupport][build.webkit.org] build step clean-git-index-lock is not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=282366">https://bugs.webkit.org/show_bug.cgi?id=282366</a>

Reviewed by Jonathan Bedard.

The build step `clean-git-index-lock` was not working as expected because on the
function `getResultSummary()` the value of the property `got_revision` is not
defined when the git command fails, so that was causing a Python exception later
when trying to extract a substring from a None value.

So the result is that when the step `clean-and-update-working-directory` failed
because git returned an error then the whole build ended with a build exception.

On top of that, the function `getResultSummary()` is not the proper place to add
more steps after the current one because this function gets executed at least 2
times at the end of the build and that results in adding the next step twice.

The proper place to inject a new step after the current one is at the `run()`
function that is guaranteed to only run once.

However, inside this `run()` function the logic was initially not working because
Buildbot runs the git command with `abandonOnFailure=True` which means that when
the command fails instead of returning a status of FAILURE it raises an exception
of type `BuildStepFailed`. Fix that by catching this exception.

On top of that set a property once the `clean-git-index-lock` runs to avoid running
it again if step `clean-and-update-working-directory` fails again, that way we
avoid and endless loop that could trigger in a corner case where git keeps failing
even after deleting the `index.lock` file.

* Tools/CISupport/build-webkit-org/steps.py:
(CheckOutSource.getResultSummary):
(CheckOutSource.run):
(CleanUpGitIndexLock.run):
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/286526@main">https://commits.webkit.org/286526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a01ed87e9189ee5995d6334f2b26a1f411854df1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55071 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80540 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27308 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3367 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59622 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17776 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39979 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46909 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22788 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25635 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68025 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82003 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67848 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/75719 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3565 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67158 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11111 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9233 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11793 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3361 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3382 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/4072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -234,20 +234,27 @@ def __init__(self, repourl='https://github.com/WebKit/WebKit.git', **kwargs):
+            rc = FAILURE
+        if rc == FAILURE:
+            if self.getProperty("cleanUpGitIndexLockAlreadyTried", False):
+                self.build.buildFinished(['Git issue'], FAILURE)
```
Do we want to retry the build as is usually the case when there is a git issue?
> `self.build.buildFinished(['Git issue, retrying build'], RETRY)`

clopez replied:
It does the retry the build, but only once, retrying it always could lead to an infinite loop.

The logic is the follow one:

1. Run step `clean-and-update-working-directory`
2. If step `clean-and-update-working-directory` fails then run step `clean-git-index-lock`
3. If step `clean-git-index-lock` fails then stop the build (as it failed to delete the `.git/index.lock` file and we don't know how to recover from that automatically)
4. If step `clean-git-index-lock` works then retry again step `clean-and-update-working-directory`.
5. If step `clean-and-update-working-directory` fails again then stop the build as we already tried to clean the `.git/index.lock` file but the step `clean-and-update-working-directory` keeps failing. That means there is an unknown issue with git that we don't know how to fix automatically and retrying again would likely cause an infinite loop.

briannafan replied:
Thanks, that makes sense to me.

Our EWS implementation actually retries the entire build after deleting `.git/index.lock` - @JonWBedard do you know if there's a benefit to retrying the entire build instead of rerunning `clean-and-update-working-directory`? If not, then this PR LGTM.

JonWBedard replied:
There isn't, @clopez is putting more thought into this that we did when in originally writing the step, I think.

briannafan commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -234,20 +234,27 @@ def __init__(self, repourl='https://github.com/WebKit/WebKit.git', **kwargs):
-        else:
+        if self.results == SUCCESS:
             return {'step': 'Cleaned and updated working directory'}
+        return {'step': 'Failed to updated working directory'}
```
typo: `updated` -> `update`

briannafan commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -268,7 +275,12 @@ def run(self):
         rc = yield super().run()
         if rc != SUCCESS:
-            self.build.buildFinished(['Git issue, retrying build'], RETRY)
+            self.build.buildFinished(['Error deleting stale .git/index.lock file'], FAILURE)
```
If deleting doesn't work, should we add a `"Please inform an admin."` line? Would this failure just persist until manual intervention?

clopez replied:
Yes, makes sense, as fixing that failure would require manual intervention

clopez committed:
[CISupport][build.webkit.org] build step clean-git-index-lock is not working
https://bugs.webkit.org/show_bug.cgi?id=282366

Reviewed by Jonathan Bedard.

The build step `clean-git-index-lock` was not working as expected because on the
function `getResultSummary()` the value of the property `got_revision` is not
defined when the git command fails, so that was causing a Python exception later
when trying to extract a substring from a None value.

So the result is that when the step `clean-and-update-working-directory` failed
because git returned an error then the whole build ended with a build exception.

On top of that, the function `getResultSummary()` is not the proper place to add
more steps after the current one because this function gets executed at least 2
times at the end of the build and that results in adding the next step twice.

The proper place to inject a new step after the current one is at the `run()`
function that is guaranteed to only run once.

However, inside this `run()` function the logic was initially not working because
Buildbot runs the git command with `abandonOnFailure=True` which means that when
the command fails instead of returning a status of FAILURE it raises an exception
of type `BuildStepFailed`. Fix that by catching this exception.

On top of that set a property once the `clean-git-index-lock` runs to avoid running
it again if step `clean-and-update-working-directory` fails again, that way we
avoid and endless loop that could trigger in a corner case where git keeps failing
even after deleting the `index.lock` file.

* Tools/CISupport/build-webkit-org/steps.py:
(CheckOutSource.getResultSummary):
(CheckOutSource.run):
(CleanUpGitIndexLock.run):
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: https://commits.webkit.org/286526@main
